### ISSUE TITLE
TCP keepalive mechanism not working

### ISSUE THREAD

**Issue by kiemlicz** [istio/istio#43233]
### Bug Description

Hello

#### Some context first

When running Istio, I'm experiencing inter-service connections RST (between in-cluster services, I'm not referring to client <> gateway connections), caused by `ISTIO_META_IDLE_TIMEOUT`.

For example (MongoDB with some Java app using it):
- during heavier load Mongo Java driver creates more connections
- as load decreases connections stay in the pool for driver to manage
- following happens:
-- driver gets the connection asynchronously from pool to be used for the upcoming query
-- Istio kills the connection due to the fact it was idle past (default) 1h
-- driver issues the query via connection that just got closed
-- "Prematurely reached end of stream" from Mongo driver is thrown

Seems like we can configure Mongo driver to have the max idle time per connection, however different applications may not have similar setting.

#### Two questions
1.
Stated here:
https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto#envoy-v3-api-field-extensions-filters-network-tcp-proxy-v3-tcpproxy-idle-timeout
If the timeout is disabled, why there is "high likehood of connection leaks"?

2.
I've deployed two simple PODs:
- server running `nc` and
- client with simple python tcp client sending/receiving strings,
having tcp keepalives configured like so:
```
import socket
import sys

HOST = sys.argv[1]
PORT = 7777  # The port used by the server

with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
    s.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
    s.setsockopt(socket.IPPROTO_TCP, socket.TCP_KEEPIDLE, 20)
    s.setsockopt(socket.IPPROTO_TCP, socket.TCP_KEEPINTVL, 15)
    s.setsockopt(socket.IPPROTO_TCP, socket.TCP_KEEPCNT, 10)
    s.connect((HOST, PORT))
    while True:
        txt = input("Enter text: ")
        s.sendall(bytes(txt, 'utf-8'))
        data = s.recv(1024)
        print(data)
```
and deployed like so:
```
---
apiVersion: v1
kind: Pod
metadata:
  name: client
  annotations:
    proxy.istio.io/config: |-
      proxyMetadata:
        ISTIO_META_IDLE_TIMEOUT: "130s"
spec:
  securityContext:
    sysctls: # not particularly relevant for this issue since using setsockopt
      - name: net.ipv4.tcp_keepalive_time
        value: "10"
      - name: net.ipv4.tcp_keepalive_intvl
        value: "3"
      - name: net.ipv4.tcp_keepalive_probes
        value: "10"
  containers:
    - name: debian
      image: localhost:5000/client:latest  # this is the aforementioned app
      imagePullPolicy: Always
      command:
        - bash  # I want to run manually
      tty: true
      stdin: true
---
apiVersion: v1
kind: Pod
metadata:
  name: server
  annotations:
    proxy.istio.io/config: |-
      proxyMetadata:
        ISTIO_META_IDLE_TIMEOUT: "130s"
spec:
  containers:
    - name: debian
      image: debian:stable
      command:
        - bash  # setup nc there
      tty: true
      stdin: true
```

When I deploy them to NS without istio injection (`k label ns some_ns istio-injection=disabled --overwrite`)
I can see the TCP keepalives send every configured interval via `setsockopt`.
On the other hand, when I deploy them with `istio-injection=enabled`, there are no keepalive messages sent
Why is that?
For example the Mongo Java driver uses keepalives to sustain the connection, however with Istio they are not working  
In other words, as a user I would not expect Istio to break the TCP keepalive

### Version

```prose
Istio:  
client version: 1.14.3
control plane version: 1.13.1
data plane version: 1.13.1 (11 proxies)
k8s:  
Client Version: v1.26.1
Kustomize Version: v4.5.7
Server Version: v1.26.1
```

### Additional Information

Attaching two pcaps (screenshot only since I don't see the option to upload file, sniffed from `any` and one port only), one without Istio, 
![image](https://user-images.githubusercontent.com/7331452/217843545-3d3d742a-e6c2-46b3-99a1-e7f27462d7c1.png)

another with Istio
![image](https://user-images.githubusercontent.com/7331452/217843753-b419143a-f7c1-421f-8ea0-5a78244e319a.png)

**Comment by kiemlicz** [comment]
No bro, not stale

**Comment by nielsole** [comment]
[Docs say](https://istio.io/latest/docs/reference/config/networking/destination-rule/#ConnectionPoolSettings-TCPSettings-TcpKeepalive):
> Default is to use the OS level configuration

Yet I don't think https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#envoy-v3-api-msg-config-core-v3-tcpkeepalive uses the OS defaults 🤔

**Comment by nielsole** [comment]
Then [envoy example section](https://www.envoyproxy.io/docs/envoy/latest/configuration/overview/examples) even says:

> It’s important to set appropriate [TCP Keep-Alive options](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#envoy-v3-api-msg-config-core-v3-tcpkeepalive) in the tcp_keepalive block.

**Comment by nielsole** [comment]
https://man7.org/linux/man-pages/man7/tcp.7.html
```
tcp_keepalive_time (integer; default: 7200; since Linux 2.2)
              The number of seconds a connection needs to be idle before
              TCP begins sending out keep-alive probes.  Keep-alives are
              sent only when the SO_KEEPALIVE socket option is enabled.
```

Confirmation on envoy behaviour to not set SO_KEEPALIVE if `tcp_keepalive: {}` is not set:
https://github.com/envoyproxy/envoy/blob/91a95cf1651bbac8c629b0903313e924e6ec6b03/test/common/upstream/cluster_manager_impl_test.cc#L5384
